File: hook-sync.js
#!/usr/bin/env node
// Hook 래퍼 — stdin에서 transcript_path 추출 후 단일 파일만 동기화
// Stop/SessionEnd 훅에서 호출됨
// 핵심: 자식 프로세스를 detach하여 부모(hook)가 죽어도 동기화 완료

const fs = require('fs');
const path = require('path');

const LOG_FILE = path.join(__dirname, 'sync.log');

function log(msg) {
  const ts = new Date().toISOString().replace('T', ' ').slice(0, 19);
  fs.appendFileSync(LOG_FILE, `[${ts}] ${msg}\n`);
}

let data = '';
process.stdin.setEncoding('utf8');
process.stdin.on('data', chunk => { data += chunk; });
process.stdin.on('end', () => {
  try {
    const hook = JSON.parse(data.trim());
    const tp = hook.transcript_path;
    const event = hook.hook_event_name || 'unknown';
    const sid = hook.session_id?.slice(0, 8) || '?';
    if (tp) {
      const { spawn } = require('child_process');
      const logFd = fs.openSync(LOG_FILE, 'a');
      const child = spawn('node', [
        path.join(__dirname, 'sync-conversations.js'),
        '--file', tp
      ], {
        detached: true,
        stdio: ['ignore', logFd, logFd],
        windowsHide: true
      });
      child.unref();
      log(`${event} → ${sid} sync started`);
    }
  } catch (err) {
    log(`ERROR: ${err.message}`);
  }
});
process.stdin.resume();

File: hook_sync.py
#!/usr/bin/env python3
"""Hook wrapper — reads stdin JSON, spawns detached sync for a single file.

Used by Stop and SessionEnd hooks. The child process is fully detached
so the parent hook can exit without waiting for sync to complete.

The child runs via _sync_runner.py which catches errors and logs them
with proper file locking to avoid concurrent write corruption.
"""

import json
import os
import subprocess
import sys
from datetime import datetime, timezone, timedelta
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
LOG_FILE = SCRIPT_DIR / 'sync.log'
RUNNER_SCRIPT = SCRIPT_DIR / '_sync_runner.py'
KST = timezone(timedelta(hours=9))

# Determine venv python path
if sys.platform == 'win32':
    PYTHON = str(SCRIPT_DIR / '.venv' / 'Scripts' / 'pythonw.exe')
else:
    PYTHON = str(SCRIPT_DIR / '.venv' / 'bin' / 'python')


def _log_ts():
    """Dual KST | UTC timestamp prefix to avoid timezone confusion when reading sync.log."""
    now_utc = datetime.now(timezone.utc)
    now_kst = now_utc.astimezone(KST)
    return f"{now_kst.strftime('%Y-%m-%d %H:%M:%S')} KST | {now_utc.strftime('%Y-%m-%d %H:%M:%S')} UTC"


def log(msg):
    with open(LOG_FILE, 'a', encoding='utf-8') as f:
        f.write(f'[{_log_ts()}] {msg}\n')


def main():
    data = sys.stdin.read().strip()
    try:
        hook = json.loads(data)
        tp = hook.get('transcript_path')
        event = hook.get('hook_event_name', 'unknown')
        sid = (hook.get('session_id') or '?')[:8]

        if tp:
            kwargs = {
                'stdin': subprocess.DEVNULL,
                'stdout': subprocess.DEVNULL,
                'stderr': subprocess.DEVNULL,
            }

            if sys.platform == 'win32':
                DETACHED = 0x00000008  # DETACHED_PROCESS
                CREATE_NO_WINDOW = 0x08000000
                kwargs['creationflags'] = DETACHED | CREATE_NO_WINDOW
            else:
                kwargs['start_new_session'] = True

            proc = subprocess.Popen(
                [PYTHON, str(RUNNER_SCRIPT), '--file', tp, '--sid', sid],
                **kwargs,
            )
            # Don't wait for child
            log(f'{event} -> {sid} sync started (pid {proc.pid})')
    except Exception as e:
        log(f'ERROR: {e}')


if __name__ == '__main__':
    main()
